codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/specs/ProtocolSpec.java

```diff
@@ -91,4 +95,67 @@ default String discoveredEndpoint(OperationModel opModel) {
                ? ".discoveredEndpoint(cachedEndpoint)\n"
                : "";
     }
+
+
+    /**
+     * For sync streaming operations, wrap request marshaller in {@link StreamingRequestMarshaller} class.
+     */
+     default CodeBlock syncStreamingMarshaller(IntermediateModel model, OperationModel opModel, ClassName marshaller) {
+        CodeBlock.Builder builder = CodeBlock
+            .builder()
+            .add("$T.builder().delegateMarshaller(new $T(protocolFactory))", StreamingRequestMarshaller.class, marshaller)
+            .add(".requestBody(requestBody)");
+
+        // Set requires length
+        if (opModel.hasRequiresLengthInInput()) {
+            builder.add(".requiresLength(true)");
+        }
+         
+        if (AuthType.V4_UNSIGNED_BODY.equals(opModel.getAuthType())) {
+            builder.add(".transferEncoding(true)");
+        }
+
+        if (model.getMetadata().supportsH2()) {
+            builder.add(".useHttp2(true)");
+        }
+
+        builder.add(".build()");
+
+        return builder.build();
+    }
+
+    default CodeBlock asyncMarshaller(IntermediateModel model, OperationModel opModel, ClassName marshaller,
+                                     String protocolFactory) {
+        if (opModel.hasStreamingInput()) {
+            return asyncStreamingMarshaller(model, opModel, marshaller, protocolFactory);
+        } else {
+            return CodeBlock.builder().add("new $T($L)", marshaller, protocolFactory).build();
+        }
+    }
+
+    default CodeBlock asyncStreamingMarshaller(IntermediateModel model, OperationModel opModel,
+                                                      ClassName marshaller, String protocolFactory) {
+        CodeBlock.Builder builder = CodeBlock
+            .builder()
+            .add("$T.builder().delegateMarshaller(new $T($L))", AsyncStreamingRequestMarshaller.class,
+                 marshaller, protocolFactory)
+            .add(".asyncRequestBody(requestBody)");
+
+        // Set requires length
+        if (opModel.hasRequiresLengthInInput()) {
```

Comment:
There are some duplicates between this method and `syncStreamingMarshaller`, can we extract them out? Otherwise it's really easy to miss either one of them in the future if we add more header logic.